commit e09cb2b6b482a76fc2200599cbd1685234aa87bc
Author: Jan Safranek <jsafrane@redhat.com>
Date:   Wed Jan 29 10:10:42 2014 +0100

    Introduce new 'TempDir' configuration option.
    
    The option can be used to set different temporary directory than /tmp.
    /tmp is unsafe and racy, admins or package maintainers might want to use
    other directory, e.g. /var/run/openlmi-storage instead.

diff --git a/doc/admin/admin-config.rst b/doc/admin/admin-config.rst
index 490ff0b..873a0db 100644
--- a/doc/admin/admin-config.rst
+++ b/doc/admin/admin-config.rst
@@ -27,26 +27,28 @@ Default configuration::
      Stderr=false
      DebugBlivet=false
 
-======= =================== ==================== ===========
-Section Option name         Default value        Desciption
-======= =================== ==================== ===========
-``CIM`` ``Namespace``       root/cimv2           Namespace where OpenLMI-Storage providers are registered.
-``CIM`` ``SystemClassName`` PG_ComputerSystem    Name of CIM_ComputerSystem class, which is used to represent the computer system. It will be used as ``SystemClassName`` property value of various classes.
-``Log`` ``Level``           ERROR                Chooses which messages are logged, either to CIMOM and (if configured) to standard error output. Available levels (sorted by severity) are:
-
-                                                    * CRITICAL
-                                                    * ERROR
-                                                    * WARNING
-                                                    * INFO
-                                                    * DEBUG
-                                                    * TRACE_WARNING
-                                                    * TRACE_INFO
-                                                    * TRACE_VERBOSE
-
-                                                 Levels below INFO (= TRACE_WARNING, TRACE_INFO and DEBUG) are useful mainly for debugging and bug reporting.
-``Log`` ``Stderr``          false                Toggles sending of log messages to standard error output of the CIMOM.
-``Log`` ``DebugBlivet``     false                Toggles logging of detailed debug messages in Blivet.
-======= =================== ==================== ===========
+=========== =================== ==================== ===========
+Section     Option name         Default value        Desciption
+=========== =================== ==================== ===========
+``CIM``     ``Namespace``       root/cimv2           Namespace where OpenLMI-Storage providers are registered.
+``CIM``     ``SystemClassName`` PG_ComputerSystem    Name of CIM_ComputerSystem class, which is used to represent the computer system. It will be used as ``SystemClassName`` property value of various classes.
+``Log``     ``Level``           ERROR                Chooses which messages are logged, either to CIMOM and (if configured) to standard error output. Available levels (sorted by severity) are:
+
+                                                        * CRITICAL
+                                                        * ERROR
+                                                        * WARNING
+                                                        * INFO
+                                                        * DEBUG
+                                                        * TRACE_WARNING
+                                                        * TRACE_INFO
+                                                        * TRACE_VERBOSE
+
+                                                     Levels below INFO (= TRACE_WARNING, TRACE_INFO and DEBUG) are useful mainly for debugging and bug reporting.
+``Log``     ``Stderr``          false                Toggles sending of log messages to standard error output of the CIMOM.
+``Log``     ``DebugBlivet``     false                Toggles logging of detailed debug messages in Blivet.
+``Storage`` ``Tempdir``         /tmp                 Path to temporary directory. The providers (usually running as root) need read/write access there.
+                                                     When SELinux or other security enhancement mechanism is used, **only** the provider should have read/write access to this directory.
+=========== =================== ==================== ===========
 
 Persistent setting
 ------------------
diff --git a/src/lmi/storage/StorageConfiguration.py b/src/lmi/storage/StorageConfiguration.py
index 3c6caef..d04e262 100644
--- a/src/lmi/storage/StorageConfiguration.py
+++ b/src/lmi/storage/StorageConfiguration.py
@@ -1,4 +1,4 @@
-# Copyright (C) 2012 Red Hat, Inc.  All rights reserved.
+# Copyright (C) 2012-2014 Red Hat, Inc.  All rights reserved.
 #
 # This library is free software; you can redistribute it and/or
 # modify it under the terms of the GNU Lesser General Public
@@ -51,6 +51,7 @@ class StorageConfiguration(BaseConfiguration):
         """ :rtype: (``dict``) Dictionary of default values. """
         defaults = BaseConfiguration.default_options().copy()
         defaults['DebugBlivet'] = 'false'
+        defaults['TempDir'] = '/tmp'
         return defaults
 
     @property
@@ -58,3 +59,7 @@ class StorageConfiguration(BaseConfiguration):
         """ Return True if blivet tracing is enabled."""
         return self.config.getboolean('Log', 'DebugBlivet')
 
+    @property
+    def temp_dir(self):
+        """ Return path to temporary directory."""
+        return self.get_safe('Storage', 'TempDir', fallback='/tmp')
diff --git a/src/lmi/storage/cimom_entry.py b/src/lmi/storage/cimom_entry.py
index ea7783d..2dee29b 100644
--- a/src/lmi/storage/cimom_entry.py
+++ b/src/lmi/storage/cimom_entry.py
@@ -1,4 +1,4 @@
-# Copyright (C) 2012 Red Hat, Inc.  All rights reserved.
+# Copyright (C) 2012-2014 Red Hat, Inc.  All rights reserved.
 #
 # This library is free software; you can redistribute it and/or
 # modify it under the terms of the GNU Lesser General Public
@@ -101,6 +101,7 @@ import lmi.providers.cmpi_logging as cmpi_logging
 from lmi.providers.ComputerSystem import get_system_name
 import blivet
 import logging
+import tempfile
 
 timer_manager = None
 indication_manager = None
@@ -140,6 +141,9 @@ def get_providers(env):
             logger.addHandler(handler)
     LOG().info("Provider init.")
 
+    # Initialize temporary directory (for blivet)
+    tempfile.tempdir = config.temp_dir
+
     # Initialize ComputerSystem.Name
     get_system_name(env)
 
diff --git a/storage.conf b/storage.conf
index 6a0c958..5993fc8 100644
--- a/storage.conf
+++ b/storage.conf
@@ -1,5 +1,12 @@
 # Sample configuration file for OpenLMI storage provider
 
+[Storage]
+
+# Directory for temporary files and directories.
+# The directory must exist when the provider starts and the provider must
+# have read+write access there. (default = /tmp)
+#TempDir = /tmp
+
 [Log]
 # Level can be set to following values:
 #   TRACE_VERBOSE, TRACE_INFO, TRACE_WARNING, DEBUG, INFO, WARNING, ERROR, CRITICAL
@@ -11,4 +18,3 @@
 
 # Do not produce Blivet debug messages (default = false)
 #DebugBlivet=false
-
